src/libcore/num/dec2flt/mod.rs

```diff
@@ -0,0 +1,228 @@
+        return Ok(x);
+    }
+    // Big32x40 is limited to 1280 bits, which translates to about 385 decimal digits.
+    // If we exceed this, perhaps while calculating `f * 10^e` in Algorithm R or Algorithm M,
+    // we'll crash. So we error out before getting too close, with a generous safety margin.
+    if max_digits > 375 {
+        return Err(PFE { __kind: FloatErrorKind::Invalid });
+    }
+    let f = digits_to_big(decimal.integral, decimal.fractional);
+
+    // Now the exponent certainly fits in 16 bit, which is used throughout the main algorithms.
+    let e = e as i16;
+    // FIXME These bounds are rather conservative. A more careful analysis of the failure modes
+    // of Bellerophon could allow using it in more cases for a massive speed up.
+    let exponent_in_range = table::MIN_E <= e && e <= table::MAX_E;
+    let value_in_range = max_digits <= T::max_normal_digits();
+    if  exponent_in_range && value_in_range {
+        Ok(algorithm::bellerophon(&f, e))
+    } else {
+        Ok(algorithm::algorithm_m(&f, e))
+    }
+}
+
+// As written, this optimizes badly (see #27130). `inline(always)` is a workaround for that.
+// There are only two call sites overall and it doesn't make code size worse.
+
+/// Strip zeros where possible, even when this requires changing the exponent
+#[inline(always)]
+fn simplify(decimal: &mut Decimal, e: &mut i64) {
+    // Trimming these zeros does not change anything but may enable the fast path (< 15 digits).
+    while decimal.integral.first() == Some(&b'0') {
+        decimal.integral = &decimal.integral[1..];
+    }
+    while decimal.fractional.last() == Some(&b'0') {
+        let end = decimal.fractional.len();
+        decimal.fractional = &decimal.fractional[..end-1];
+    }
+    // Simplify numbers of the form 0.0...x and x...0.0, adjusting the exponent accordingly.
+    // This may not always be a win (possibly pushes some numbers out of the fast path), but it
+    // simplifies other parts significantly (notably, approximating the magnitude of the value).
+    if decimal.integral.is_empty() {
+        while decimal.fractional.first() == Some(&b'0') {
+            decimal.fractional = &decimal.fractional[1..];
+            *e -= 1;
+        }
+    } else if decimal.fractional.is_empty() {
+        while decimal.integral.last() == Some(&b'0') {
+            let end = decimal.integral.len();
+            decimal.integral = &decimal.integral[..end-1];
+            *e += 1;
+        }
+    }
+}
+
+/// Detect obvious overflows and underflows without even looking at the decimal digits.
+fn trivial_cases<T: RawFloat>(decimal: &Decimal, e: i64) -> Option<T> {
+    // There were zeros but they were stripped by simplify()
+    if decimal.integral.is_empty() && decimal.fractional.is_empty() {
+        return Some(T::zero());
+    }
+    // This is a crude approximation of ceil(log10(the real value)).
+    let max_place = e + decimal.integral.len() as i64;
```

Comment:
Might want to explicitly note why the math involving `e` in this file can't overflow; it's a bit subtle.